politeiawww/user/mysql/mysql.go

```diff
@@ -0,0 +1,1028 @@
+	sync.RWMutex
+
+	shutdown       bool                            // Backend is shutdown
+	userDB         *sql.DB                         // Database context
+	encryptionKey  *[32]byte                       // Data at rest encryption key
+	pluginSettings map[string][]user.PluginSetting // [pluginID][]PluginSettings
+}
+
+func ctxWithTimeout() (context.Context, func()) {
+	return context.WithTimeout(context.Background(), connTimeout)
+}
+
+func (m *mysql) isShutdown() bool {
+	m.RLock()
+	defer m.RUnlock()
+
+	return m.shutdown
+}
+
+// encrypt encrypts the provided data with the mysql encryption key. The
+// encrypted blob is prefixed with an sbox header which encodes the provided
+// version. The read lock is taken despite the encryption key being a static
+// value because the encryption key is zeroed out on shutdown, which causes
+// race conditions to be reported when the golang race detector is used.
+//
+// This function must be called without the lock held.
+func (m *mysql) encrypt(version uint32, b []byte) ([]byte, error) {
+	m.RLock()
+	defer m.RUnlock()
+
+	return sbox.Encrypt(version, m.encryptionKey, b)
+}
+
+// decrypt decrypts the provided packed blob using the mysql encryption
+// key. The read lock is taken despite the encryption key being a static value
+// because the encryption key is zeroed out on shutdown, which causes race
+// conditions to be reported when the golang race detector is used.
+//
+// This function must be called without the lock held.
+func (m *mysql) decrypt(b []byte) ([]byte, uint32, error) {
+	m.RLock()
+	defer m.RUnlock()
+
+	return sbox.Decrypt(m.encryptionKey, b)
+}
+
+// setPaywallAddressIndex updates the paywall address index record in the
+// key-value store.
+//
+// This function must be called using a transaction.
+func setPaywallAddressIndex(ctx context.Context, tx *sql.Tx, index uint64) error {
+	b := make([]byte, 8)
+	binary.LittleEndian.PutUint64(b, index)
+	_, err := tx.ExecContext(ctx,
+		"UPDATE key_value SET v = $1 WHERE k = $2", b, keyPaywallAddressIndex)
+	if err != nil {
+		return fmt.Errorf("update paywallet index error: %v", err)
+	}
+	return nil
+}
+
+// userNew creates a new user the database.  The userID and paywall address
+// index are set before the user record is inserted into the database.
+//
+// This function must be called using a transaction.
+func (m *mysql) userNew(ctx context.Context, tx *sql.Tx, u user.User) (*uuid.UUID, error) {
+	// Set user paywall address index.
+	var index uint64
+	var dbIndex []byte
+	err := tx.QueryRowContext(ctx, "SELECT v FROM key_value WHERE k = $1",
+		keyPaywallAddressIndex).Scan(&dbIndex)
+	switch err {
+	// No errors, use database index.
+	case nil:
+		index = binary.LittleEndian.Uint64(dbIndex) + 1
+	// No rows found error; Index wasn't initiated in table yet, default to zero.
+	case sql.ErrNoRows:
+		index = 0
+	// All other errors.
+	default:
+		return nil, fmt.Errorf("find paywall index: %v", err)
+	}
+
+	u.PaywallAddressIndex = index
```

Comment:
Can we add a debug statement here that prints the paywall address index.  I want to be able to verify that its incrementing properly when I'm testing it.

Reply:
surething